volumina/tiling/cache.py

```diff
@@ -0,0 +1,299 @@
+import threading
+import warnings
+import logging
+from collections import OrderedDict, defaultdict
+
+import numpy
+from PyQt5.QtWidgets import QGraphicsItem
+
+logger = logging.getLogger()
+
+__all__ = ["TilesCache"]
+
+
+class CachePolicy:
+    def __init__(self, size):
+        self._validate_size(size)
+        self._size = size
+        self._subsribers = []
+
+    @property
+    def size(self):
+        return self._size
+
+    def _validate_size(self, value):
+        if not isinstance(value, int) or value <= 0:
+            raise ValueError("size should non negative integer")
+
+    def set_size(self, value):
+        if self._size == value:
+            return
+
+        self._validate_size(value)
+
+        self._size = value
+
+        for sub in self._subsribers:
+            sub()
+
+    def subscribe(self, fn):
+        self._subsribers.append(fn)
+
+    def unsubsribe(self, fn):
+        try:
+            self._subsribers.remove(fn)
+        except ValueError:
+            pass
+
+
+class MultiCacheException(Exception):
```

Comment:
Rename the exception class to something more descriptive like `DuplicateKeyError`?

Can also move the error message to `__init__`:
```python
def __init__(self, key):
    super().__init__(f"Duplicate key {key}")
```